pkg/webhook/workloadspread/validating/workloadspread_validation.go

```diff
@@ -0,0 +1,253 @@
+						allErrs = append(allErrs, field.Invalid(fldPath.Index(i).Child("requiredNodeSelectorTerm"), subset.RequiredNodeSelectorTerm, fmt.Sprintf("Convert_v1_NodeSelectorTerm_To_core_NodeSelectorTerm failed: %v", err)))
+					} else {
+						allErrs = append(allErrs, corevalidation.ValidateNodeSelectorTerm(*coreNodeSelectorTerm, fldPath.Index(i).Child("requiredNodeSelectorTerm"))...)
+					}
+				}
+
+				if subset.PreferredNodeSelectorTerms != nil {
+					corePreferredSchedulingTerms := make([]core.PreferredSchedulingTerm, 0, len(subset.PreferredNodeSelectorTerms))
+					for i, term := range subset.PreferredNodeSelectorTerms {
+						corePreferredSchedulingTerm := &core.PreferredSchedulingTerm{}
+						if err := corev1.Convert_v1_PreferredSchedulingTerm_To_core_PreferredSchedulingTerm(term.DeepCopy(), corePreferredSchedulingTerm, nil); err != nil {
+							allErrs = append(allErrs, field.Invalid(fldPath.Index(i).Child("preferredSchedulingTerms"), subset.PreferredNodeSelectorTerms, fmt.Sprintf("Convert_v1_PreferredSchedulingTerm_To_core_PreferredSchedulingTerm failed: %v", err)))
+						} else {
+							corePreferredSchedulingTerms = append(corePreferredSchedulingTerms, *corePreferredSchedulingTerm)
+						}
+					}
+
+					allErrs = append(allErrs, corevalidation.ValidatePreferredSchedulingTerms(corePreferredSchedulingTerms, fldPath.Index(i).Child("preferredSchedulingTerms"))...)
+				}
+			}
+
+			if subset.Tolerations != nil {
+				var coreTolerations []core.Toleration
+				for i, toleration := range subset.Tolerations {
+					coreToleration := &core.Toleration{}
+					if err := corev1.Convert_v1_Toleration_To_core_Toleration(&toleration, coreToleration, nil); err != nil {
+						allErrs = append(allErrs, field.Invalid(fldPath.Index(i).Child("tolerations"), subset.Tolerations, fmt.Sprintf("Convert_v1_Toleration_To_core_Toleration failed: %v", err)))
+					} else {
+						coreTolerations = append(coreTolerations, *coreToleration)
+					}
+				}
+				allErrs = append(allErrs, corevalidation.ValidateTolerations(coreTolerations, fldPath.Index(i).Child("tolerations"))...)
+			}
+
+			if subset.MaxReplicas != nil {
+				if i < len(subsets)-1 {
+					subsetMaxReplicas, err := intstr.GetValueFromIntOrPercent(subset.MaxReplicas, 100, true)
+					if err != nil || subsetMaxReplicas < 0 {
+						allErrs = append(allErrs, field.Invalid(fldPath.Index(i).Child("maxReplicas"), subset.MaxReplicas, "maxReplicas is not valid for subset"))
+					}
+					if subset.MaxReplicas.Type == intstr.String && subsetMaxReplicas > 100 {
+						allErrs = append(allErrs, field.Invalid(fldPath.Index(i).Child("maxReplicas"), subset.MaxReplicas, "maxReplicas exceeds 100% for subset"))
+					}
+				} else {
+					// validate maxReplicas of the default subset
+					allErrs = append(allErrs, field.Invalid(fldPath.Index(i).Child("maxReplicas"), subset.MaxReplicas, "maxReplicas of default subset must be nil"))
```

Comment:
Do this mean if WorkloadSpread CR has only one subset, then its replica number has to be nil, which means the CR does not affect any scheduling decision? 

If so, should we enforce that a WorkloadSpread CR has to have at least two subsets?

Reply:
One subset also affects scheduling decisions because the webhook injects the subset topology rules into the Pod, and the maxReplicas is nil, which means there is no limit for replica numbers in the subset.
All in all, a subset can distribute the workload in a fixed zone.

Reply:
This makes no difference comparing to applying node selector and tolerations in workload Pod Spec directly, isn't it?

Reply:
Indeed, one subset has no meaning for workload spread, and should has at least two subset. I will adjust the validate rules to enforce CR at least two subset.

Reply:
@Fei-Guo 
This is considered because there is a moment of inconsistency between spec and status (the user adjusts the spec and then the controller takes some time to adjust the status), and at this point it is hoped that there will be a default subset to give the webhook the ability to do the injection.

Now this is not really appropriate, so webhook will determine whether the spec.generation and status.generation are equal when injecting, if not, etc. temporarily reject the request, and then let the workload controller retry to ensure that the webhook injection when the spec and status is consistent. So it is not necessary to set the last subset to nil.